### PR TITLE
Stabilize `uniform_paths`

### LINKED ISSUE #53130: (Modules) Tracking issue for Picking a Module Path System variant
[Relative Paths in Rust 2018]: https://internals.rust-lang.org/t/relative-paths-in-rust-2018/7883

This is a sub-tracking issue for the RFC "Clarify and streamline paths and visibility" (rust-lang/rfcs#2126)
as well as the internals post [Relative Paths in Rust 2018].

The issue deals with revamping paths and making a decision about whether we should pick:
+ The `uniform_paths` variant.
    Proposed in [Relative Paths in Rust 2018].
+ The `anchored_use_paths` variant.
    You have to use `self::`, `super::`, and `crate::`, or an external crate name explicitly....
    See rust-lang/rfcs#2126).

- [x] Implement all the things.
   - [x] `uniform_paths`, https://github.com/rust-lang/rust/pull/52923
   - [x] `anchored_use_paths`
- [x] Make both variants available to test in Edition Preview 2 with appropriate documentation on how to test.
- [x] Adjust documentation ([see instructions on forge][doc-guide])
- [ ] Pick a variant
- [ ] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

**Unresolved questions:**

- [ ] Which variant do we pick?

### LINKED ISSUE #55618: Stabilize uniform paths on Rust 2018
@rfcbot fcp merge
cc #53130

This issue is to track the discussion of the following proposal:

> I propose we stabilize the `uniform_paths` feature in Rust 2018, eliminating the `anchored_paths` variant. I further propose we backport this stabilization to version 1.31 beta, so that all stable versions of Rust 2018 have this feature.

We've taken informal polls about this question [in a paper document][paper]. Most of the lang team has favored uniform_paths, I and @cramertj were the only two members who initially favored anchored_paths. Yesterday, we discussed this in the lang team video meeting; in the meeting we sort of crystallized the various pros and cons of the two variants, but didn't reach a consensus.

In the time since the meeting, I've come around to thinking that stabilizing on uniform_paths is probably the best choice. I imagine @cramertj has not changed his mind, and though he said in the meeting he wouldn't block a decision, I hope we can use this FCP period to see if there's any agreeable changes that would make uniform_paths more palatable to him.

I've also written [a blog post][blog] summarizing the meeting discussion and how my position has evolved, which provides more context on this decision.

I also want to make sure we continue to get community feedback as we reach this final decision! Please contribute new thoughts you have to the discussion. However, we will hopefully make this decision within the next few weeks, and certain proposals are out of scope. Any third proposal which is not completely backwards compatible with the current behavior of paths in Rust 2018 is impossible to ship in Rust 2018 at this point. Variations on anchored or uniform paths that *are* backwards compatible may be considered, but that also seems unlikely.

[paper]: https://paper.dropbox.com/doc/Lang-team-stance-on-modules-cvcNJv9qnNUeWXMcvLGzs
[blog]: https://boats.gitlab.io/blog/post/anchored-uniform/

### LINKED ISSUE #56326: edition idioms: Incorrect suggestion for removing `extern crate ignore;`
Given this code:

```
$ cat ignore.rs
#![crate_type = "lib"]

pub fn foo() {}
$ cat foo.rs
#![warn(rust_2018_idioms)]

extern crate ignore;

fn main() {
    ignore::foo();
}
$ rustc +beta ignore.rs
$ rustc +beta foo.rs --extern ignore=./libignore.rlib --edition 2018
warning: `extern crate` is not idiomatic in the new edition
 --> foo.rs:3:1
  |
3 | extern crate ignore;
  | ^^^^^^^^^^^^^^^^^^^^ help: convert it to a `use`
  |
note: lint level defined here
 --> foo.rs:1:9
  |
1 | #![warn(rust_2018_idioms)]
  |         ^^^^^^^^^^^^^^^^
  = note: #[warn(unused_extern_crates)] implied by #[warn(rust_2018_idioms)]
```

If applied, though, the suggestion yields:

```
$ rustc +beta foo.rs --extern ignore=./libignore.rlib --edition 2018
error[E0658]: imports can only refer to extern crate names passed with `--extern` on stable channel (see issue #53130)
 --> foo.rs:3:5
  |
3 | use ignore;
  |     ^^^^^^ not an extern crate passed with `--extern`
  |
note: this import refers to the built-in attribute imported here
 --> foo.rs:3:5
  |
3 | use ignore;
  |     ^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0658`.
```

First reported at https://github.com/rust-lang/cargo/issues/6353 along with a [suggestion of how to fix it](https://github.com/rust-lang/cargo/issues/6353#issuecomment-442536847)

### LINKED ISSUE #56398: Build error with extern crates after upgrading to 2018 edition
I took my project here and converted it to the 2018 edition: https://github.com/elliottslaughter/integrity-checker

The flow I used was:

```
git clone https://github.com/elliottslaughter/integrity-checker
cd integrity-checker
cargo +beta fix --edition # no warnings on first try, one change
git commit -a -m 'Prep for edition.'
# go to Cargo.toml and set edition keyword
cargo +beta build
```

But even though `cargo fix` completes without warnings, the build is giving me:

```
   Compiling integrity-checker v0.1.0 (/Users/elliott/Programming/Rust/integrity-checker)                                                   
error[E0658]: imports can only refer to extern crate names passed with `--extern` on stable channel (see issue #53130)                      
 --> src/error.rs:2:5                                                                                                                       
  |                                                                                                                                         
2 | use ignore;                                                                                                                             
  |     ^^^^^^ not an extern crate passed with `--extern`                                                                                   
  |                                                                                                                                         
note: this import refers to the built-in attribute imported here                                                                            
 --> src/error.rs:2:5                                                                                                                       
  |                                                                                                                                         
2 | use ignore;                                                                                                                             
  |     ^^^^^^                                                                                                                              
                                                                                                                                            
error: aborting due to previous error                                                                                                       
                                                                                                                                            
For more information about this error, try `rustc --explain E0658`.                                                                         
error: Could not compile `integrity-checker`.                                                                                               

To learn more, run the command again with --verbose.
```

For posterity I've saved the result of having applied cargo fix here:

https://github.com/elliottslaughter/integrity-checker/tree/rust-2018

```
$ rustc +beta --version
rustc 1.31.0-beta.20 (8940af3ce 2018-11-30)
$ cargo +beta --version
cargo 1.31.0-beta (339d9f9c8 2018-11-16)
```

### LINKED ISSUE #56417: Stabilize uniform paths on Rust 2018 (technical details)
https://github.com/rust-lang/rust/issues/55618 is the high-level and high-volume thread on whether we want uniform import paths in the language in general or not.
This is a more focused issue about how exactly stabilization will happen (assuming it will happen in general).

First, regarding timing of the stabilization.
I propose to test uniform paths for 4-5 weeks starting from Dec 07 (Rust 2018 release, 1.31 stable), and then backport their stabilization on 1.32 beta if everything is good.

Second, regarding sub-features and partial stabilization.
Imports `use NAME` or `use NAME::...` in the uniform path model can refer to various entities, not all of which may be expected or were discussed.
Here's the list:
- Items defined in named or unnamed modules (`mod m { struct NAME; }`, `fn f() { struct NAME; }`).
No known issues to resolve before stabilization.
- Macros from other crates imported with `#[macro_use] extern crate ...`, for example `use panic` from the standard library.
No known issues to resolve before stabilization.
- Extern crate names from extern prelude, for example `use std` or `use regex`.
No known issues to resolve before stabilization.
- Names from the standard library prelude, for example `use Vec`.
No known issues to resolve before stabilization.
- Built-in types, for example `use u8`.
No known issues to resolve before stabilization.
- Built-in macros, for example `use env`.
Currently an error due to some (fixable) implementation details of built-in macros.
No known issues to resolve before stabilization (after the error is removed).
- Macros defined with `macro_rules!` in the same crate, e.g. `macro_rules! mac {()=>()}    use mac as pac`
Unresolved question: what visibility to attach to these macros, in other words - how far can they be reexported with `pub use`?
Proposal: treat `#[macro_export] macro_rules! { ... }` as `pub`, treat other `macro_rules! { ... }` as `pub(crate)`.
Motivation: 1) `#[macro_export]` are indeed visible from other crates, 2) non-`#[macro_export]` macros themselves are indeed potentially visible from the whole crate, it depends on the containing module whether to actually let them out or not (similarly to public items in private modules and their potential reexports).
- Built-in attributes, for example `use inline`.
Issue: even if `inline` is reimported under some other name, e.g. `use inline as my_inline`, `my_inline` won't be treated as `inline` by the compiler. Even later stages of the compiler work with attributes at token level, not using resolution results. That's not good in general, ideally attributes should be lowered into some semantic form somewhere around AST -> HIR conversion.
This means a compatibility hazard, for example `#[my_repr(D)] fn f() {}` would be accepted and ignored if attributes are treated syntactically (assuming `use repr as my_repr`), but would be an error if attributes are treated semantically based on their resolution.
On the other hand, if `use builtin_attr` is still feature-gated, then things like `use proc_macro` (or `use ignore` as [recently reported](https://github.com/rust-lang/rust/issues/56398)) will be feature gated as well (`use` imports in all namespaces, and `proc_macro` is not only a crate, but also a built-in attribute).
Proposal: Allow imports of built-in attributes, but prohibit actually using names imported this way in attribute positions.
- Derive helper attributes registered by derive macros, for example `use serde` attribute registered by `Serialize` macro.
Not fully implemented, so imports can never refer to them.
Issue (once fully implemented): similarly to built-in attributes, derive helpers reimported under other name will be unrecognizable by their respective proc macros, because proc macros work at token level.
Proposal: Allow imports of derive helper attributes, but prohibit actually using names imported this way in attribute positions.
- "Tool modules" in tool attributes, like `rusfmt` in `rustfmt::skip`.
Issue: similarly to built-in attributes, tool modules reimported under other name will be unrecognizable by their respective tools, because tools work at token level.
Proposal: Allow imports of tool modules, but prohibit actually using names imported this way in attribute paths.

### LINKED ISSUE #56821: Unhelpful error message for missing `crate` in uses import
This problem occurs with the new path rules in edition 2018. If the `crate` keyword from a uses import is missing, the current error message is not very helpful. [For example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5fdfbecf7b9683a6a652fe918a21b651):

```rust
mod bla {
    pub struct Bla;
}

use bla::Bla;
```

returns on stable, beta and nightly the following error message:

```
error[E0658]: imports can only refer to extern crate names passed with `--extern` on stable channel (see issue #53130)
 --> src/main.rs:9:5
  |
5 | / mod bla {
6 | |     pub struct Bla;
7 | | }
  | |_- not an extern crate passed with `--extern`
8 | 
9 |   use bla::Bla;
  |       ^^^
  |
note: this import refers to the module defined here
 --> src/main.rs:5:1
  |
5 | / mod bla {
6 | |     pub struct Bla;
7 | | }
  | |_^

warning: unused import: `bla::Bla`
 --> src/main.rs:9:5
  |
9 | use bla::Bla;
  |     ^^^^^^^^
  |
  = note: #[warn(unused_imports)] on by default
```
The compiler looks for a crate `bla` but doesn't find one, therefore the error message. I'm not sure, how this relates to the upcoming decision regarding the efforts for *path clarity*. But I stumbled over this several times now, out of habit...

(If this issue is accepted, I would like to improve it :) I would probably need mentoring though.)

### LINKED ISSUE #57252: 2018 edition fails to import modules for binary projects
Hello Rust developers,

My project is a binary project:

```shell
├── Cargo.lock
├── Cargo.toml
├── src
    ├── a.rs
    ├── main.rs
    └── z.rs
```

Without `edition = "2018"`, my imports work:

```rust
mod a;
mod z;
```

With `edition = "2018"`, above imports don't work. Prefixing `crate::a` or `self::a` doesn't work either.

Error message is:

```shell
error[E0658]: imports can only refer to extern crate names passed with `--extern` on stable channel (see issue #53130)
  --> src/main.rs:13:5
   |
1  | mod a;
   | ------ not an extern crate passed with `--extern`
```

I stripped down the error message at line `#13`, it's just `use a::something;`.

I think it's a bug, may I ask if you could verify?

Thank you,

### LINKED ISSUE #57422: edition compatbility lint: fails to handle macro and crate with same name
The following code compiles without warnings in 2015:

```rust
#![warn(rust_2018_compatibility)]
#[macro_use]
extern crate serde_json;
extern crate json;

pub mod m {
    use json;
    pub fn f() {
        json::parse("").unwrap();
    }
}

pub fn f() {
    println!("{:?}", json!({}));
}
```

In 2018, it fails to compile with the following error:
```
error[E0658]: imports can only refer to extern crate names passed with `--extern` on stable channel (see issue #53130)
 --> src/lib.rs:7:9
  |
2 | #[macro_use]
  | ------------ not an extern crate passed with `--extern`
...
7 |     use json;
  |         ^^^^
  |
  = help: add #![feature(uniform_paths)] to the crate attributes to enable
note: this import refers to the macro imported here
 --> src/lib.rs:2:1
  |
2 | #[macro_use]
  | ^^^^^^^^^^^^
```

Notice the `json` macro conflicts with the `json` crate.

### PR DESCRIPTION
**Pull request by petrochenkov**
Address all the things described in https://github.com/rust-lang/rust/issues/56417.

Assign visibilities to `macro_rules` items - `pub` to `macro_export`-ed macros and `pub(crate)` to non-exported macros, these visibilities determine how far these macros can be reexported with `use`.

Prohibit use of reexported inert attributes and "tool modules", after renaming (e.g. `use inline as imported_inline`) their respective tools and even compiler passes won't be able to recognize and properly check them.

Also turn use of uniform paths in 2015 macros into an error, I'd prefer to neither remove nor stabilize them right away because I still want to make some experiments in this area (uniform path fallback was added to 2015 macros used on 2018 edition in https://github.com/rust-lang/rust/pull/56053#issuecomment-441405140).

UPDATE: The last commit also stabilizes the feature `uniform_paths`.

Closes https://github.com/rust-lang/rust/issues/53130
Closes https://github.com/rust-lang/rust/issues/55618
Closes https://github.com/rust-lang/rust/issues/56326
Closes https://github.com/rust-lang/rust/issues/56398
Closes https://github.com/rust-lang/rust/issues/56417
Closes https://github.com/rust-lang/rust/issues/56821
Closes https://github.com/rust-lang/rust/issues/57252
Closes https://github.com/rust-lang/rust/issues/57422

### CODE REVIEW

**Comment by petrochenkov** [comment]
@Centril 
I've added one more commit actually stabilizing the feature.
Could you start an FCP?

**Comment by Centril** [comment]
@petrochenkov 

> @Centril
> I've added one more commit actually stabilizing the feature.
> Could you start an FCP?

I've been meaning to read this and the other issues / descriptions you wrote but I've been a bit pressed for time. Would you mind waiting a bit (~a day) before I start FCP?

**Comment by petrochenkov** [comment]
@Centril 
Sure, a day is okay, we have 3 weeks until the release.

File: src/test/ui/rust-2018/uniform-paths/prelude-fail-2.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// edition:2018
+
+#![feature(uniform_paths)]
+
+// Built-in attribute
+use inline as imported_inline;
+
+#[imported_inline] //~ ERROR cannot use a built-in attribute through an import
```
Comment (Centril) [inline src/test/ui/rust-2018/uniform-paths/prelude-fail-2.rs]: Is there a test for derive helper attrs somewhere?

**Review by Centril — APPROVED** [review]
Tests, and from what I can see implementation, looks good.

Reassigning for further technical review to:
r? @nikomatsakis

**Comment by Centril** [comment]
~A day later...

> + Proposal: treat #[macro_export] macro_rules! { ... } as pub, treat other macro_rules! { ... } as pub(crate).
> + Proposal: Allow imports of built-in attributes, but prohibit actually using names imported this way in attribute positions.
> + Proposal: Allow imports of derive helper attributes, but prohibit actually using names imported this way in attribute positions.
> + Proposal: Allow imports of tool modules, but prohibit actually using names imported this way in attribute paths.
> + Some of the errors are artificial and can be relaxed backward-compatibly in the future, making import paths even closer to non-import paths.

This seems like the right incremental balance for the time being. 👍 

All in all, let's:

@rfcbot merge

----------------------

> [Known deviations of macro paths and 2018 import paths from the "uniform path" model](https://github.com/rust-lang/rust/issues/56413)
>> There's still some future-proofing in place turning imports referring to local variables and generic parameters into errors - 22a13ef, but that future proofing has two holes.

These 2 holes seem quite pathological... but how much work work would it be to fix them?

File: src/test/ui/rust-2018/uniform-paths/prelude-fail-2.rs (hunk 1 above)
Comment (petrochenkov) [reply src/test/ui/rust-2018/uniform-paths/prelude-fail-2.rs]: Derive helper resolution is generally under-implemented and they can never be imported right now (beside other issues).

They are taken into account only when we are resolving 1) an attribute which is 2) placed on the same item with derive.

**Comment by nikomatsakis** [comment]
I have a question: From the comments, it seemed like there was some "concern" about stabilizing this in time for the next release, is there any particular reason for that urgency?

**Comment by nikomatsakis** [comment]
(Not saying I'm *opposed*, just asking)

**Comment by petrochenkov** [comment]
@nikomatsakis 
If this is about my concerns in https://github.com/rust-lang/rust/issues/55618#issuecomment-437205772, then it was about stabilizing it in 1.31, literally days after landing the re-implementation.
Now the edition was released, many project migrated and no insurmountable difficulties during migration were reported, so I'm confident enough that we can ship this.

Otherwise the decision to stabilize was already taken in https://github.com/rust-lang/rust/issues/55618.
It would be ideal if we could do that for 1.31, since import reform is one of the key features of 2018 edition, but that didn't work out. 1.32 is good enough too.
Delaying this further until 1.33 would probably be not so good.

**Comment by Centril** [comment]
To expand on reasons for urgency and why we should do it in 1.32, I think that the more we delay, the more likely it is that `uniform_paths` will not be the idiomatic way for some time in the 2018 edition.

**Comment by pnkfelix** [comment]
@rfcbot reviewed

**Comment by nikomatsakis** [comment]
OK, checked.

**Review by nikomatsakis — APPROVED** [review]
r=me

**Comment by Mark-Simulacrum** [comment]
Tentatively beta-accepting. Discussed on Discord and decided that this was of sufficient importance (due to bugfixes) to try and land, modulo a successful crater run.

**Comment by petrochenkov** [comment]
Beta backport - https://github.com/rust-lang/rust/pull/57483.

**Comment by petrochenkov** [comment]
@bors p=1
(Needs to get into the upcoming beta.)

**Comment by petrochenkov** [comment]
>The job exceeded the maximum time limit for jobs, and has been terminated.

Looks spurious.
@bors retry

**Pull request merged**